### ISSUE TITLE
unboxing different types/enums

### ISSUE THREAD

**Issue by tcurdt** [JohnSundell/Unbox#103]
I have to different kind of types in the json and I am not sure how to handle the unboxing:

    "foo": {
        "open": {
          "from": "2016-08-14T13:00:00",
          "to": "2016-08-14T18:00:00"
        }
      }

and

    "foo": {
        "close": {
          "from": "2016-08-14T13:00:00",
          "to": "2016-08-14T18:00:00"
        }
      }

I guess the Swift representation could look like either

    enum SpecialType {
        case Open(from:NSDate, until:NSDate)
        case Closed(from:NSDate, until:NSDate)
    }

    struct Foo {
      let foo: SpecialType
    }

or

    enum SpecialType {
        case Open
        case Closed
    }

    struct Foo {
      let from: NSDate
      let until: NSDate
      let type: SpecialType
    }

Any advice?

**Comment by clayellis** [comment]
Hi, @tcurdt! Welcome to Unbox! Glad you're here. Consider the following. It doesn't make use of `enum`s, but I think it makes a little more sense. If you see this and don't like the implementation let me know and I can help you create the "enum" strategy. Cheers!

### Foo
```swift 
struct Foo {
  let open: OpenClosed? 
  let closed: OpenClosed?
}

extension Foo: Unboxable {
  init(unboxer: Unboxer) {
    open = unboxer.unbox("open")
    closed = unboxer.unbox("close")
  }
}
```

### OpenClosed
```swift
struct OpenClosed {
  let from: NSDate
  let until: NSDate
}

extension OpenClosed: Unboxable {
  init(unboxer: Unboxer) {
    let formatter = NSDateFormatter()
    formatter.dateFormat = "YYYY-MM-DD'T'HH:MM:SS"
    from = unboxer.unbox("from", formatter: formatter)
    until = unboxer.unbox("to", formatter: formatter)
  }
}
```

With this strategy, `Foo.open` and `Foo.closed` are both optional and you could check which case it is by using an `if let` statement.

```swift
let foo: Foo = try? Unbox(json)
if let open = foo.open {
  // The store is open!
} else if let closed = foo.closed {
  // Out for lunch...
}
```

**Comment by tcurdt** [comment]
Thanks for the help and the warm welcome, @clayellis :)

I came up with a work around along the lines of:

    enum SpecialType {
        case Open(from:NSDate, until:NSDate)
        case Closed(from:NSDate, until:NSDate)
    }

    let openFrom: NSDate? = unboxer.unbox("value.open.from", isKeyPath: true, formatter: dateFormatter)
    let openUntil: NSDate? = unboxer.unbox("value.open.to", isKeyPath: true, formatter: dateFormatter)

    let closedFrom: NSDate? = unboxer.unbox("value.close.from", isKeyPath: true, formatter: dateFormatter)
    let closedUntil: NSDate? = unboxer.unbox("value.close.to", isKeyPath: true, formatter: dateFormatter)

    if openFrom != nil && openUntil != nil {
        foo = Open(from:openFrom, until: openUntil)
    } else {
        foo = Closed(from:closedFrom, until: closedUntil)
    }

...which seems a little bit closer to the model I was hoping for.

Unfortunately I could not work out a way to handle the optionals/non-match case nicely. It's not like I could throw an error. I guess it will have to do something like

    enum SpecialType {
        case Open(from:NSDate, until:NSDate)
        case Closed(from:NSDate, until:NSDate)
        case Undefined
    }

So far `unbox` has been a real pleasure to use - this case is a bit ugly to deal with though.

**Comment by clayellis** [comment]
@JohnSundell do you see any other ways to improve unboxing this case? If not, I think this issue can be closed.

**Comment by tcurdt** [comment]
I know the change is probably too big - but just being able to somehow error out of `init(unboxer: Unboxer)` would be great.

**Comment by tcurdt** [comment]
I found `unboxer.failForKey(key: "")` but it does not help with `Returns from initializer without initializing all stores properties`.

The only way I see that works around the problem with non-optionals is to use `throws` as in changing the signature and error handling to:

```
required init(unboxer: Unboxer) throws {
}
```

**Comment by JohnSundell** [comment]
This has been fixed in Unbox 2.0, where you can now throw from an initializer.
